Upgrade 9.0 updates (#800)
elastic/docs-content: deploy-manage/upgrade/deployment-or-cluster.md

The procedures you follow to upgrade depend on whether you’ve installed Elastic components using Elastic-managed infrastructure or self-managed infrastructure.

If you’re running Elastic-managed infrastructure, your options are to:

* [Upgrade on {{ech}}](/deploy-manage/upgrade/deployment-or-cluster/upgrade-on-ech.md)
* Upgrade on {{serverless-full}} (updates are automatic and require no user management)

If you’re running your own self-managed infrastructure — either on-prem or on public cloud — your options are to:

* [Upgrade the {{stack}}](/deploy-manage/upgrade/deployment-or-cluster/self-managed.md) (upgrade each component individually)
* [Upgrade on {{ece}} (ECE)](/deploy-manage/upgrade/deployment-or-cluster/upgrade-on-ece.md)
* [Upgrade on {{eck}} (ECK)](/deploy-manage/upgrade/deployment-or-cluster/upgrade-on-eck.md)

## Prepare to upgrade [prepare-to-upgrade]

Before you upgrade Elastic, it's important to take some preparation steps. These steps vary based on your current version. 

:::{important}

> Upgrading from a release candidate build, such as 9.0.0-rc1 or 9.0.0-rc2, is not supported. Pre-releases should only be used for testing in a temporary environment.

Reviewer: ```suggestion
Upgrading from a release candidate build, such as 9.0.0-rc1, is not supported. Pre-releases should only be used for testing in a temporary environment.
```
No rc2 this time

:::

## Prepare to upgrade from 8.x [prepare-upgrade-from-8.x]

> To upgrade to 9.0 from 8.17 or earlier, you must first upgrade to the latest patch version of 8.18. This enables you to use the [Upgrade Assistant](prepare-to-upgrade/upgrade-assistant.md) to identify and resolve issues, reindex indices created before 8.0, and then perform a rolling upgrade. Upgrading to 8.18 before upgrading to 9.x is required even if you opt to do a full-cluster restart of your {{es}} cluster. If you're running a pre-8.x version, you might need to perform multiple upgrades or a full-cluster restart to get to 8.18 to prepare to upgrade to 9.0.

Reviewer: ```suggestion
To upgrade to 9.0 from 8.17 or earlier, you must first upgrade to the latest patch version of 8.18. This enables you to use the [Upgrade Assistant](prepare-to-upgrade/upgrade-assistant.md) to identify and resolve issues, reindex indices created before 8.0, and then perform a rolling upgrade. Upgrading to 8.18 before upgrading to 9.0 is required even if you opt to do a full-cluster restart of your {{es}} cluster. If you're running a pre-8.x version, you might need to perform multiple upgrades or a full-cluster restart to get to 8.18 to prepare to upgrade to 9.0.
```
Since 8.18 is only for 9.0, we can be specific.
When 9.1 comes out, it will be 8.19 to 9.1+

Alternatively, you can create a new 9.0 deployment and reindex from remote. For more information, refer to [Reindex to upgrade](#reindex-to-upgrade).

:::{note}
{{beats}} and {{ls}} 8.18 are compatible with {{es}} 9.x to give you flexibility in scheduling the upgrade. {{es}} 8.x clients are also compatible with 9.x and use [REST API compatibility](elasticsearch://reference/elasticsearch/rest-apis/compatibility.md) by default to help ensure compatibility between 8.x clients and the 9.x {{es}} server. 
:::

With the exception of serverless, the following recommendations are best practices for all deployment methods.

1. Run the [Upgrade Assistant](prepare-to-upgrade/upgrade-assistant.md) to prepare for your upgrade from 8.18 to 9.0. The Upgrade Assistant identifies deprecated settings, and guides you through resolving issues, and reindexing data streams and indices created before 8.0. 

    :::{note}

>     Please be aware that depending on your setup, if your indices change due to reindexing, you might need to change alerts, transforms, or other code that was targeting the old index.

Reviewer 2: ```suggestion
    Depending on your setup, reindexing can change your indices, and you may need to update alerts, transforms, or other code targeting the old index.
```

    :::

> 2. Ensure you have a current [snapshot](/deploy-manage/tools/snapshot-and-restore/create-snapshots.md) before making configuration changes or reindexing. 

Reviewer 2: ```suggestion
2. Before you change configurations or reindex, make sure you have a current [snapshot](/deploy-manage/tools/snapshot-and-restore/create-snapshots.md). 
```

    :::{tip}
    Tip: From version 8.3, snapshots are generally available as simple archives. Use the [archive functionality](/deploy-manage/upgrade/deployment-or-cluster/reading-indices-from-older-elasticsearch-versions.md) to search snapshots as old as version 5.0 without the need of an old {{es}} cluster. This ensures that data you store in {{es}} doesn’t have an end of life and is still accessible when you upgrade, without requiring a reindex process.
    :::

    You must resolve all critical issues before proceeding with the upgrade. If you make any additional changes, take a new snapshot to back up your data. 

3. Review the deprecation logs from the Upgrade Assistant to determine if your applications are using features that are not supported or behave differently in 9.x. 

> 4. Major version upgrades can include breaking changes that require you to take additional steps to ensure that your applications behave as expected after the upgrade. Review all breaking changes for each product you use to review more information about changes that could affect your application. Make sure you test against the new version before upgrading existing deployments.

Reviewer 2: ```suggestion
4. Major version upgrades can include breaking changes that require additional steps to ensure your applications function as expected. Review the breaking changes for each product you use to learn more about potential impacts on your application. Make sure you test with the new version before upgrading existing deployments.
```

> 5. Make the recommended changes to ensure that your clients continue to operate as expected after the upgrade. 

Reviewer 2: ```suggestion
5. To ensure your clients continue to operate as expected after the upgrade, make the recommended changes. 
```

    :::{note}

>     As a temporary solution, you can submit requests to 9.x using the 8.x syntax with the REST API compatibility mode. While this enables you to submit requests that use the old syntax, it does not guarantee the same behavior. REST API compatibility should be a bridge to smooth out the upgrade process, not a long term strategy. For more information about how to best leverage REST API compatibility during an upgrade, refer to [REST API compatibility](elasticsearch://reference/elasticsearch/rest-apis/compatibility.md). 

Reviewer 2: ```suggestion
    As a temporary solution, use the 8.x syntax to submit requests to 9.0.0 with REST API compatibility mode. While this allows you to submit requests using the old syntax, it doesn’t guarantee the same behavior. REST API compatibility should serve as a bridge during the upgrade, not a long-term solution. For more details on how to effectively use REST API compatibility during an upgrade, refer to [REST API compatibility](elasticsearch://reference/elasticsearch/rest-apis/compatibility.md). 
```

    :::

6. If you use any {{es}} plugins, make sure there is a version of each plugin that is compatible with the {{es}} version you're upgrading to.

> 7. We recommend creating a 9.0 test deployment and test the upgrade in an isolated environment before upgrading your production deployment. Ensure that both your test and production environments have the same settings.

Reviewer 2: ```suggestion
7. Before upgrading your production deployment, create a 9.0.0 test deployment and test the upgrade in an isolated environment. Make sure the test and production environments use the same settings.
```

    :::{note}
    If you use {{ccs}}, note that 9.0+ can only search remote clusters running the  previous minor version, the same version, or a newer minor version in the same major version. For more information, refer to [Cross-cluster search](../../solutions/search/cross-cluster-search.md).

    If you use {{ccr}}, a cluster that contains follower indices must run the same or newer (compatible) version as the remote cluster. For more information and to view the version compatibility matrix, refer to [Cross cluster replication](/deploy-manage/tools/cross-cluster-replication.md). You can view your remote clusters from **Stack Management > Remote Clusters**.
    ::::

9. Consider closing {{ml}} jobs before you start the upgrade process. While {{ml}} jobs can continue to run during a rolling upgrade, it increases the overhead on the cluster during the upgrade process.

10. If you have any anomaly detection result indices `.ml-anomalies-*` that were created in {{es}} 7.x, they must be reindexed, marked as read-only, or deleted before upgrading to 9.x. To learn how to do this, refer to [Anomaly detection results migration](#anomaly-migration). 

> 11. If you have any transform destination indices that were created in {{es}} 7.x, they must be reset, reindexed, or deleted before upgrading to 9.x. To learn how to do this, refer to [Transform destination indices migration](#transform-migration). 

Reviewer 2: ```suggestion
11. If you have transform destination indices created in {{es}} 7.x, reset, reindex, or delete them before you upgrade to 9.0.0. For more information, refer to [Transform destination indices migration](#transform-migration). 
```

> ## Reindex to upgrade (optional) [reindex-to-upgrade]

Reviewer 2: ```suggestion
## Reindex to upgrade [reindex-to-upgrade]
```

> To create a new 9.0 deployment and reindex from remote:

Reviewer 2: ```suggestion
Optionally create a 9.0.0 deployment and reindex from remote.
```

> 1. Provision an additional deployment running 9.0.

Reviewer 2: ```suggestion
1. Provision an additional deployment running 9.0.0.
```

To set the index as read-only, add the write block to the index:

```
PUT .ml-anomalies-custom-example/_block/write
```

Indices created in {{es}} 7.x that have a write block will not raise a critical deprecation warning.
:::

:::{dropdown} Deleting anomaly result indices
If an index contains results of the jobs that are no longer required. To list all jobs that stored results in an index, use the terms aggregation:

```json
GET .ml-anomalies-custom-example/_search
{
  "size": 0,
  "aggs": {
    "job_ids": {
      "terms": {
        "field": "job_id",
        "size": 100
      }
    }
  }
}
```

The jobs can be deleted in the UI. After the last job is deleted, the index will be deleted as well.
:::

## Transform destination indices migration [transform-migration]

> =======

Reviewer 3: Not sure if this is still WIP, but just flagging that this section is blank.

Author: Yep, a WIP!
